ethpm/manifest.py

```diff
@@ -0,0 +1,257 @@
+
+    :build_dependencies:        [optional]
+        Dict[str, str] with the package name pointing to a content-
+        addressable URI which resolves to the package's manifest.
+
+    :allow_extra_meta_fields:   [optional]
+        Flag that must be explicitly set to true if user wants to allow
+        fields in the meta object that are not listed in the spec.
+    """
+    validate_manifest_version(manifest_version)
+    validate_package_name(package_name)
+    validate_package_version(version)
+    manifest = {
+        "manifest_version": manifest_version,
+        "package_name": package_name,
+        "version": version,
+    }
+    build_fns = (
+        build_meta(meta, allow_extra_meta_fields),
+        build_sources(compiled_json_output_path),
+        build_contract_types(compiled_json_output_path, contract_types),
+        build_build_dependencies(build_dependencies),
+    )
+    built_manifest = cytoolz.pipe(manifest, *build_fns)
+    return built_manifest
+
+
+#
+# Meta
+#
+
+
+@cytoolz.curry
+def build_meta(
+    meta: Dict[str, Any], allow_extra_meta_fields: bool, manifest: Dict[str, Any]
+) -> Dict[str, Any]:
+    """
+    Validate meta object keys and values, and attach object to manifest if valid.
+    """
+
+    if meta:
+        validate_meta_object(meta, allow_extra_meta_fields)
+        new_manifest = cytoolz.assoc(manifest, "meta", meta)
+        return new_manifest
+    return manifest
+
+
+def validate_meta_object(meta: Dict[str, Any], allow_extra_meta_fields: bool) -> None:
+    """
+    Validates that every key is one of `META_FIELDS` and has a value of the expected type.
+    """
+    for key, value in meta.items():
+        if key in META_FIELDS:
+            if type(value) is not META_FIELDS[key]:
+                raise ValidationError(
+                    "Values for {0} are expected to have the type {1}, instead got {2}.".format(
+                        key, META_FIELDS[key], type(value)
+                    )
+                )
+            continue
+        if allow_extra_meta_fields:
+            if key[:2] != "x-":
+                raise ValidationError(
+                    "Undefined meta fields need to begin with 'x-', "
+                    "{0} is not a valid undefined meta field.".format(key)
+                )
+        else:
+            raise ValidationError(
+                "{0} is not a permitted meta field. To allow undefined fields, "
+                "set `allow_extra_meta_fields` to True.".format(key)
+            )
+
+
+#
+# Sources
+#
+
+
+@cytoolz.curry
+def build_sources(
+    compiled_json_output_path: Path, manifest: Dict[str, Any]
+) -> Dict[str, Any]:
+    """
+    Generate all source keys/values found in the linked compiled json output, and
+    attach them to the manifest.
+    """
+    if compiled_json_output_path:
+        source_data = generate_all_sources(compiled_json_output_path)
+        new_manifest = cytoolz.assoc(manifest, "sources", source_data)
```

Comment:
These might be more legible without the variable and just returning directly from here.